### ISSUE TITLE
URL Routing Support

### ISSUE THREAD

**Issue by Archmonger** [idom-team/idom#569]
### Current Situation

Currently, URL routes cannot be resolved within IDOM.

### Proposed Changes

Add support for URL routing within IDOM core.

### Implementation Details

Add `React-Router` into IDOM core. This likely will be in the form of a `route` tag.

```python
return html.route("my_path/to/thing", html.div("hello world"))
```

**Comment by Archmonger** [comment]
Examples have been added on potential styling patterns.

There's a very high chance we pick the react style, but just want to make sure options are listed.

**Comment by Archmonger** [comment]
I've updated this with a deep dive into what React-Router API endpoints need to be replicated for MVP URL routing support.

**Comment by rmorshea** [comment]
I've done a little bit of thinking on this here and there and at the moment I'm leaning towards the idea of making this a separate package (e.g. `idom-router`). I'm also imagining that the implementation might rely on [Web Components](https://developer.mozilla.org/en-US/docs/Web/Web_Components).

**Comment by Archmonger** [comment]
If it can be a separate package then it should be. 
If you believe the current infrastructure is flexible enough to support URL routing outside of core, then I'd say to go down that path.

**Comment by rmorshea** [comment]
Agreed. Especially given this is how react does it. With that said, without charges to core I'm not sure how clean the implementation will be. Regardless, I'm thinking I hack a prototype and see how bad it is before I go making changes in core before I really groc how routing works.

**Comment by rmorshea** [comment]
I've created a draft implementation here: https://github.com/idom-team/idom-router

I would discourage using it until I've had the time to flesh it out with more tests and documentation. The interface is subject to change (I haven't looked too closely at how React Router does things yet), but for the moment, usage looks like:

```python
from idom import component, html, run
from idom.backend import default

from idom_router import Route, Link

Router = bind(default)

@component
def Sample():
    return Router(
        Route("/", Link({"id": "root"}, "to a", to="/a")),
        Route("/a", Link({"id": "a"}, "to b", to="/b")),
        Route("/b", Link({"id": "b"}, "to c", to="/c")),
        Route("/c", Link({"id": "c"}, "to default", to="/default")),
        Route("/*", html.h1({"id": "default"}, "Default")),
    )

run(Sample)
```

Route patterns follow standard glob rules (see `fnmatch` module for details) unless an `re.Pattern` is provided.

**Comment by Archmonger** [comment]
I'll give the source a quick glance.

Let me know when it's more fleshed out and I can do a formal review if needed.

I haven't personally used react router but I watched a handful of YouTube videos on it when drafting this issue.

**Comment by Archmonger** [comment]
I'm not entirely sure what you're trying to do with the third and fourth positional arguments (`"to a", to="/a"`), but everything else about the overall design for `Router` and `Route` seems solid.

I would rename `Router` to `BrowserRouter` to be consistent with `react-router`.

**Comment by Archmonger** [comment]
Also see this review comment, since it's related to `idom-router` now:

- https://github.com/idom-team/django-idom/pull/77#discussion_r906653261

**Comment by Archmonger** [comment]
@rmorshea I can't recall if I asked this before or not but...

Is there a technical reason why `module_from_template` can't be used for `react-router`?

**Comment by Archmonger** [comment]
This issue should be transferred to [idom-team/idom-router](https://github.com/idom-team/idom-router)
